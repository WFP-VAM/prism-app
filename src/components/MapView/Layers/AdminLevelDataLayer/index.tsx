import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { get } from 'lodash';
import { GeoJSONLayer } from 'react-mapbox-gl';
import {
  AdminLevelDataLayerProps,
  BoundaryLayerProps,
  LayerKey,
} from '../../../../config/types';
import {
  LayerData,
  loadLayerData,
} from '../../../../context/layers/layer-data';
import {
  layerDataSelector,
  mapSelector,
} from '../../../../context/mapStateSlice/selectors';
import { addLayer, removeLayer } from '../../../../context/mapStateSlice';
import { addPopupData } from '../../../../context/tooltipStateSlice';
import { getFeatureInfoPropsData } from '../../utils';
import {
  getBoundaryLayers,
  getBoundaryLayerSingleton,
  LayerDefinitions,
} from '../../../../config/utils';
import { addNotification } from '../../../../context/notificationStateSlice';
import { isLayerOnView } from '../../../../utils/map-utils';
import { getRoundedData } from '../../../../utils/data-utils';
import { useSafeTranslation } from '../../../../i18n';
import { fillPaintData } from '../styles';

function AdminLevelDataLayers({ layer }: { layer: AdminLevelDataLayerProps }) {
  const dispatch = useDispatch();
  const map = useSelector(mapSelector);
  const boundaryId = layer.boundary || getBoundaryLayerSingleton().id;

  const layerData = useSelector(layerDataSelector(layer.id)) as
    | LayerData<AdminLevelDataLayerProps>
    | undefined;
  const { data } = layerData || {};
  const { features } = data || {};
  const { t } = useSafeTranslation();

  useEffect(() => {
    // before loading layer check if it has unique boundary?
    const boundaryLayers = getBoundaryLayers();
    const boundaryLayer = LayerDefinitions[
      boundaryId as LayerKey
    ] as BoundaryLayerProps;

    if ('boundary' in layer) {
      if (Object.keys(LayerDefinitions).includes(boundaryId)) {
        boundaryLayers.map(l => dispatch(removeLayer(l)));
        dispatch(addLayer({ ...boundaryLayer, isPrimary: true }));

        // load unique boundary only once
        // to avoid double loading which proven to be performance issue
        if (!isLayerOnView(map, boundaryId)) {
          dispatch(loadLayerData({ layer: boundaryLayer }));
        }
      } else {
        dispatch(
          addNotification({
            message: `Invalid unique boundary: ${boundaryId} for ${layer.id}`,
            type: 'error',
          }),
        );
      }
    }
    if (!features) {
      dispatch(loadLayerData({ layer }));
    }
  }, [dispatch, features, layer, boundaryId, map]);

  if (!features) {
    return null;
  }

  if (!isLayerOnView(map, boundaryId)) {
    return null;
  }

  return (
    <GeoJSONLayer
      before={`layer-${boundaryId}-line`}
      id={`layer-${layer.id}`}
      data={features}
      fillPaint={fillPaintData(layer)}
      fillOnClick={async (evt: any) => {
        // by default add `data_field` to the tooltip
        dispatch(
          addPopupData({
            [layer.title]: {
              data: getRoundedData(get(evt.features[0], 'properties.data'), t),
              coordinates: evt.lngLat,
            },
          }),
        );
        // then add feature_info_props as extra fields to the tooltip
        dispatch(
          addPopupData(
            getFeatureInfoPropsData(layer.featureInfoProps || {}, evt),
          ),
        );
      }}
    />
  );
}

export default AdminLevelDataLayers;
