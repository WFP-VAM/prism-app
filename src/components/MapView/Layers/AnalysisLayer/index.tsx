import React from 'react';
import { get } from 'lodash';
import { GeoJSONLayer } from 'react-mapbox-gl';
import * as MapboxGL from 'mapbox-gl';
import { useDispatch, useSelector } from 'react-redux';
import { addPopupData } from '../../../../context/tooltipStateSlice';
import {
  analysisResultSelector,
  isAnalysisLayerActiveSelector,
} from '../../../../context/analysisResultStateSlice';
import { legendToStops } from '../layer-utils';
import { LegendDefinition } from '../../../../config/types';

function AnalysisLayer() {
  // TODO maybe in the future we can try add this to LayerType so we don't need exclusive code in Legends and MapView to make this display correctly
  // Currently it is quite difficult due to how JSON focused the typing is. We would have to refactor it to also accept layers generated on-the-spot
  const analysisData = useSelector(analysisResultSelector);
  const isAnalysisLayerActive = useSelector(isAnalysisLayerActiveSelector);

  const dispatch = useDispatch();

  if (!analysisData || !isAnalysisLayerActive) {
    return null;
  }

  // We use the legend values from the baseline layer
  function fillPaintData(legend: LegendDefinition): MapboxGL.FillPaint {
    return {
      'fill-opacity': 0.3,
      'fill-color': {
        property: 'data',
        stops: legendToStops(legend),
        type: 'interval',
      },
    };
  }

  return (
    <GeoJSONLayer
      below="boundaries"
      data={analysisData.featureCollection}
      fillPaint={fillPaintData(analysisData.legend)}
      fillOnClick={(evt: any) => {
        const coordinates = evt.lngLat;
        dispatch(
          addPopupData({
            [analysisData.getBaselineLayer().title]: {
              data: get(evt.features[0], 'properties.data', 'No Data'),
              coordinates,
            },
            [`${analysisData.getHazardLayer().title} (${
              analysisData.statistic
            })`]: {
              data: get(
                evt.features[0],
                ['properties', analysisData.statistic],
                'No Data',
              ),
              coordinates,
            },
          }),
        );
      }}
    />
  );
}

export default AnalysisLayer;
