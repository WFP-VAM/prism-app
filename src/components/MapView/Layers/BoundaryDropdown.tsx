import {
  CircularProgress,
  FormControl,
  InputAdornment,
  InputLabel,
  ListSubheader,
  makeStyles,
  MenuItem,
  Select,
  styled,
  TextField,
  TextFieldProps,
  Theme,
  Typography,
  useMediaQuery,
  withStyles,
} from '@material-ui/core';
import { last, sortBy } from 'lodash';
import React, { forwardRef, ReactNode, useEffect, useState } from 'react';
import { Feature } from 'geojson';
import { multiPolygon, MultiPolygon, Polygon } from '@turf/helpers';
import bbox from '@turf/bbox';
import bboxPolygon from '@turf/bbox-polygon';
import union from '@turf/union';
import { useDispatch, useSelector } from 'react-redux';
import { Search, CenterFocusWeak } from '@material-ui/icons';
import { BoundaryLayerProps } from '../../../config/types';
import { appConfig, enableNavigationDropdown } from '../../../config';
import {
  getSelectedBoundaries,
  setIsSelectionMode,
  setSelectedBoundaries as setSelectedBoundariesRedux,
} from '../../../context/mapSelectionLayerStateSlice';
import { getBoundaryLayerSingleton } from '../../../config/utils';
import { Extent } from './raster-utils';
import {
  layerDataSelector,
  mapSelector,
} from '../../../context/mapStateSlice/selectors';
import { LayerData } from '../../../context/layers/layer-data';

const boundaryLayer = getBoundaryLayerSingleton();
const ClickableListSubheader = styled(ListSubheader)(({ theme }) => ({
  // Override the default list subheader style to make it clickable
  pointerEvents: 'inherit',
  cursor: 'pointer',
  '&:hover': {
    backgroundColor: theme.palette.grey[100],
  },
}));
const useStyles = makeStyles((theme: Theme) => ({
  searchField: {
    '&>div': {
      color: 'black',
    },
  },
  dropdownMenu: {
    backgroundColor: theme.palette.primary.main,
    color: 'white',
    display: 'inline-flex',
    marginTop: '1em',
    padding: theme.spacing(1, 2.66),
    borderRadius: '4px',
    alignItems: 'center',
    boxShadow: theme.shadows[2],
  },
  formControl: {
    width: '140px',
    marginLeft: '10px',
  },
  icon: {
    alignSelf: 'end',
    marginBottom: '0.4em',
  },
}));
const TIMEOUT_ANIMATION_DELAY = 10;
const SearchField = forwardRef(
  (
    {
      // important this isn't called `value` since this would confuse <Select/>
      // the main purpose of wrapping this text-field is for this very purpose.
      search,
      setSearch,
    }: {
      search: string;
      setSearch: (val: string) => void;
    },
    ref: TextFieldProps['ref'],
  ) => {
    const styles = useStyles();
    return (
      <TextField
        ref={ref}
        onKeyDown={e => e.stopPropagation()}
        className={styles.searchField}
        value={search}
        onChange={e => {
          setSearch(e.target.value);
          // when something is selected, and the user tries to search, this field deselects for some reason,
          // thus reselect on change. Important to capture target as it's null inside timeout.
          const { target } = e;
          setTimeout(() => {
            target.focus();
          }, TIMEOUT_ANIMATION_DELAY);
        }}
        InputProps={{
          startAdornment: (
            <InputAdornment position="end">
              <Search />
            </InputAdornment>
          ),
        }}
      />
    );
  },
);

/**
 * Converts the boundary layer data into a list of options for the dropdown
 * grouped by admin level 2, with individual sections under admin level 3.
 */
function getCategories(
  data: LayerData<BoundaryLayerProps>['data'],
  search: string,
) {
  if (!boundaryLayer.adminLevelNames.length) {
    console.error(
      'Boundary layer has no admin level names. Cannot generate categories.',
    );
    return [];
  }

  // Make categories based off the level of all boundaries
  return sortBy(
    data.features
      .reduce<
        Array<{
          title: string;
          children: { value: string; label: string }[];
        }>
      >((ret, feature) => {
        const parentCategory =
          feature.properties?.[boundaryLayer.adminLevelNames[0]];
        const label =
          feature.properties?.[last(boundaryLayer.adminLevelNames)!];
        const code = feature.properties?.[boundaryLayer.adminCode];
        if (!label || !code || !parentCategory) {
          return ret;
        }
        // filter via search
        const searchIncludes = (field: string) =>
          field.toLowerCase().includes(search.toLowerCase());
        if (
          search &&
          !searchIncludes(label) &&
          !searchIncludes(code) &&
          !searchIncludes(parentCategory)
        ) {
          return ret;
        }
        // add to categories if exists
        const category = ret.find(c => c.title === parentCategory);
        if (category) {
          // eslint-disable-next-line fp/no-mutating-methods
          category.children.push({ value: code, label });
        } else {
          return [
            ...ret,
            {
              title: parentCategory,
              children: [{ value: code, label }],
            },
          ];
        }
        return ret;
      }, [])
      .map(category => ({
        ...category,
        // sort children by label
        children: sortBy(category.children, 'label'),
      })),
    // then finally sort categories.
    'title',
  );
}

/**
 * This component allows you to give the user the ability to select several admin_boundary cells.
 * This component also syncs with the map automatically, allowing users to select cells by clicking the map.
 * Selection mode is automatically toggled based off this component's lifecycle.
 */
function SimpleBoundaryDropdown({
  selectedBoundaries,
  setSelectedBoundaries,
  labelMessage,
  selectAll,
  ...rest
}: BoundaryDropdownProps) {
  const [search, setSearch] = useState('');

  const boundaryLayerData = useSelector(layerDataSelector(boundaryLayer.id)) as
    | LayerData<BoundaryLayerProps>
    | undefined;
  const { data } = boundaryLayerData || {};
  if (!data) {
    return <CircularProgress size={24} color="secondary" />;
  }
  const categories = getCategories(data, search);
  const allChildren = categories.flatMap(c => c.children);
  const selectOrDeselectAll = (e: React.MouseEvent) => {
    e.preventDefault();
    if (selectedBoundaries.length > 0) {
      setSelectedBoundaries([]);
    } else {
      setSelectedBoundaries(allChildren.map(({ value }) => value));
    }
  };
  // It's important for this to be another component, since the Select component
  // acts on the `value` prop, which we need to hide from <Select/> since this isn't a menu item.
  return (
    <FormControl {...rest}>
      <InputLabel>{labelMessage}</InputLabel>
      <Select
        multiple
        onClose={() => {
          // empty search so that component shows correct options
          // otherwise, we would only show selected options which satisfy the search
          setTimeout(() => setSearch(''), TIMEOUT_ANIMATION_DELAY);
        }}
        value={selectedBoundaries}
        // Current mui version does not have SelectChangeEvent<T>. Using any instead.
        onChange={(e: any) => {
          // do nothing if value is invalid
          // This happens when you click list subheadings.
          if (
            !Array.isArray(e.target.value) ||
            e.target.value.includes(undefined)
          ) {
            return;
          }

          setSelectedBoundaries(
            Array.isArray(e.target.value) ? e.target.value : [],
            e.shiftKey,
          );
        }}
      >
        <SearchField search={search} setSearch={setSearch} />
        {!search && selectAll && (
          <MenuItem onClick={selectOrDeselectAll}>
            {selectedBoundaries.length === 0 ? 'Select All' : 'Deselect All'}
          </MenuItem>
        )}
        {search && allChildren.length === 0 && (
          <MenuItem disabled>No Results</MenuItem>
        )}
        {categories.reduce<ReactNode[]>(
          // map wouldn't work here because <Select> doesn't support <Fragment> with keys, so we need one array
          (components, category) => [
            ...components,
            // don't add list subheader if there are no categories.
            boundaryLayer.adminLevelNames.length > 1 ? (
              <ClickableListSubheader
                key={category.title}
                onClick={e => {
                  e.preventDefault();
                  // if all children are selected, deselect all. Otherwise select all
                  const categoryValues = category.children.map(c => c.value);
                  const areAllChildrenSelected =
                    selectedBoundaries.filter(val =>
                      categoryValues.includes(val),
                    ).length === categoryValues.length;

                  setSelectedBoundaries(
                    areAllChildrenSelected
                      ? selectedBoundaries.filter(
                          val => !categoryValues.includes(val),
                        )
                      : [...selectedBoundaries, ...categoryValues],
                    true,
                  );
                }}
              >
                <Typography variant="body2" color="primary">
                  {category.title}
                </Typography>
              </ClickableListSubheader>
            ) : null,
            ...category.children.map(({ label, value }) => (
              <MenuItem key={value} value={value}>
                {label}
              </MenuItem>
            )),
          ],
          [],
        )}
      </Select>
    </FormControl>
  );
}

interface BoundaryDropdownProps {
  className: string;
  selectedBoundaries: string[];
  setSelectedBoundaries: (boundaries: string[], appendMany?: boolean) => void;
  labelMessage?: string;
  selectAll: boolean;
}

/**
 * A HOC (higher order component) that connects the boundary dropdown to redux state
 */
function BoundaryDropdown({
  ...rest
}: Omit<
  BoundaryDropdownProps,
  'selectedBoundaries' | 'setSelectedBoundaries' | 'labelMessage' | 'selectAll'
>) {
  const isMobile = useMediaQuery((theme: Theme) =>
    theme.breakpoints.only('xs'),
  );
  const labelMessage = `${isMobile ? 'Tap' : 'Click'} the map to select`;

  const dispatch = useDispatch();
  const selectedBoundaries = useSelector(getSelectedBoundaries);
  // toggle the selection mode as this component is created and destroyed.
  // (users can only click the map to select while this component is visible)
  useEffect(() => {
    dispatch(setIsSelectionMode(true));
    return () => {
      dispatch(setIsSelectionMode(false));
    };
  }, [dispatch]);
  return (
    <SimpleBoundaryDropdown
      {...rest}
      selectedBoundaries={selectedBoundaries}
      setSelectedBoundaries={newSelectedBoundaries => {
        dispatch(setSelectedBoundariesRedux(newSelectedBoundaries));
      }}
      labelMessage={labelMessage}
      selectAll
    />
  );
}

export const ButtonStyleBoundaryDropdown = withStyles(() => ({
  root: {
    '& label': {
      textTransform: 'uppercase',
      letterSpacing: '3px',
      fontSize: '11px',
      position: 'absolute',
      top: '-13px',
    },
    '& svg': { color: 'white', fontSize: '1.25rem' },
    '& .MuiInput-root': { margin: 0 },
    '& .MuiInputLabel-shrink': { display: 'none' },
  },
}))(SimpleBoundaryDropdown);

export const GotoBoundaryDropdown = () => {
  const map = useSelector(mapSelector);

  const [boundaries, setBoundaries] = useState<string[]>([]);

  const boundaryLayerData = useSelector(layerDataSelector(boundaryLayer.id)) as
    | LayerData<BoundaryLayerProps>
    | undefined;
  const { data } = boundaryLayerData || {};

  const {
    map: { latitude, longitude, zoom },
  } = appConfig;

  const styles = useStyles();

  if (!data || !map || !enableNavigationDropdown) {
    return null;
  }

  return (
    <div className={styles.dropdownMenu}>
      <CenterFocusWeak fontSize="small" className={styles.icon} />
      <ButtonStyleBoundaryDropdown
        selectedBoundaries={boundaries}
        selectAll={false}
        labelMessage="Go to"
        className={styles.formControl}
        setSelectedBoundaries={(newSelectedBoundaries, appendMany) => {
          setBoundaries(() => {
            if (appendMany === true) {
              return newSelectedBoundaries;
            }

            return newSelectedBoundaries.slice(-1);
          });

          if (newSelectedBoundaries.length === 0) {
            map.flyTo({ center: { lng: longitude, lat: latitude }, zoom });

            return;
          }

          const geometries = data.features
            .filter(f =>
              newSelectedBoundaries.includes(
                f.properties && f.properties[boundaryLayer.adminCode],
              ),
            )
            .filter(f => f.geometry.type === 'MultiPolygon')
            .map(f => f.geometry as MultiPolygon);

          const bboxes = geometries.map(geom => {
            const turfObj = multiPolygon(geom.coordinates);
            const geomBbox = bbox(turfObj);

            return geomBbox;
          });

          const bboxPolygons = bboxes.map(box => bboxPolygon(box));
          const unionBbox = bboxPolygons.reduce((unionPolygon, polygon) => {
            const unionObj = union(unionPolygon, polygon);
            if (!unionObj) {
              return unionPolygon;
            }
            return unionObj as Feature<Polygon>;
          }, bboxPolygons[0]);

          map.fitBounds(bbox(unionBbox) as Extent, {
            padding: 30,
          });
        }}
      />
    </div>
  );
};

export default BoundaryDropdown;
