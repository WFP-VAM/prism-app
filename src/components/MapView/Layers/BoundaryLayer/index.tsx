import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { get } from 'lodash';
import { GeoJSONLayer } from 'react-mapbox-gl';
import * as MapboxGL from 'mapbox-gl';
import { showPopup } from '../../../../context/tooltipStateSlice';
import { BoundaryLayerProps } from '../../../../config/types';
import { layerDataSelector } from '../../../../context/mapStateSlice';
import { LayerData } from '../../../../context/layers/layer-data';

/**
 * To activate fillOnClick option, we "fill in"
 * polygons with opacity 0.
 */
const fillPaint: MapboxGL.FillPaint = {
  'fill-opacity': 0,
};

function onToggleHover(cursor: string, targetMap: MapboxGL.Map) {
  // eslint-disable-next-line no-param-reassign, fp/no-mutation
  targetMap.getCanvas().style.cursor = cursor;
}

const getLinePaintOptions: (
  layer: BoundaryLayerProps,
) => MapboxGL.LinePaint = layer => {
  return {
    'line-color': 'grey',
    'line-width': 1,
    'line-opacity': layer.opacity,
  };
};

function BoundaryLayer({ layer }: { layer: BoundaryLayerProps }) {
  const dispatch = useDispatch();
  const boundaryLayer = useSelector(layerDataSelector(layer.id)) as
    | LayerData<BoundaryLayerProps>
    | undefined;
  const { data } = boundaryLayer || {};
  if (!data) {
    return null; // boundary layer hasn't loaded yet. We load it on init inside MapView. We can't load it here since its a dependency of other layers.
  }
  return (
    <GeoJSONLayer
      id="boundaries"
      data={data}
      fillPaint={fillPaint}
      linePaint={getLinePaintOptions(layer)}
      fillOnMouseEnter={(evt: any) => onToggleHover('pointer', evt.target)}
      fillOnMouseLeave={(evt: any) => onToggleHover('', evt.target)}
      fillOnClick={(evt: any) => {
        const coordinates = evt.lngLat;
        const locationName = get(evt.features[0], 'properties.ADM1_EN', '')
          .concat(', ')
          .concat(get(evt.features[0], 'properties.ADM2_EN', ''));
        dispatch(showPopup({ coordinates, locationName }));
      }}
    />
  );
}

export default BoundaryLayer;
