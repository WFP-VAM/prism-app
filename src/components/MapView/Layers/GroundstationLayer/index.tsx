import React, { useEffect } from 'react';
import { get } from 'lodash';
import { GeoJSONLayer } from 'react-mapbox-gl';
import * as MapboxGL from 'mapbox-gl';
import { useSelector, useDispatch } from 'react-redux';
import { legendToStops } from '../layer-utils';
import { GroundstationLayerProps } from '../../../../config/types';
import { layerDataSelector } from '../../../../context/mapStateSlice';
import { addPopupData } from '../../../../context/tooltipStateSlice';
import {
  LayerData,
  loadLayerData,
} from '../../../../context/layers/layer-data';

function GroundstationLayers({ layer }: { layer: GroundstationLayerProps }) {
  const layerData = useSelector(layerDataSelector(layer.id)) as
    | LayerData<GroundstationLayerProps>
    | undefined;
  const dispatch = useDispatch();

  const { data } = layerData || {};

  useEffect(() => {
    if (!data) {
      dispatch(loadLayerData({ layer }));
    }
  }, [data, dispatch, layer]);

  if (!data) {
    return null;
  }

  const circleLayout: MapboxGL.CircleLayout = { visibility: 'visible' };
  const circlePaint: MapboxGL.CirclePaint = {
    'circle-color': {
      property: layer.measure,
      stops: legendToStops(layer.legend),
    },
  };

  return (
    <GeoJSONLayer
      below="boundaries"
      data={data}
      circleLayout={circleLayout}
      circlePaint={circlePaint}
      circleOnClick={(evt: any) => {
        dispatch(
          addPopupData({
            [layer.title]: {
              data: get(
                evt.features[0],
                `properties.${layer.measure}`,
                'No Data',
              ),
              coordinates: evt.lngLat,
            },
          }),
        );
      }}
    />
  );
}

export default GroundstationLayers;
