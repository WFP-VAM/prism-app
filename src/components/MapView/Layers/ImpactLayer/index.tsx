import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { GeoJSONLayer } from 'react-mapbox-gl';
import { FillPaint, LinePaint } from 'mapbox-gl';
import { get } from 'lodash';
import { createStyles, withStyles, WithStyles, Theme } from '@material-ui/core';
import { getExtent, Extent } from '../raster-utils';
import { legendToStops } from '../layer-utils';
import { ImpactLayerProps } from '../../../../config/types';
import {
  LayerDefinitions,
  getBoundaryLayerSingleton,
} from '../../../../config/utils';
import {
  LayerData,
  loadLayerData,
} from '../../../../context/layers/layer-data';

import { addPopupData } from '../../../../context/tooltipStateSlice';
import {
  dateRangeSelector,
  layerDataSelector,
  mapSelector,
} from '../../../../context/mapStateSlice/selectors';
import { getFeatureInfoPropsData, getRoundedData } from '../../utils';

const linePaint: LinePaint = {
  'line-color': 'grey',
  'line-width': 1,
  'line-opacity': 0.3,
};

function getHazardData(evt: any, operation: string) {
  const data = get(evt.features[0].properties, operation || 'median', null);

  return getRoundedData(data);
}

const ImpactLayer = ({ classes, layer }: ComponentProps) => {
  const map = useSelector(mapSelector);
  const { startDate: selectedDate } = useSelector(dateRangeSelector);
  const { data, date } =
    (useSelector(layerDataSelector(layer.id, selectedDate)) as LayerData<
      ImpactLayerProps
    >) || {};
  const dispatch = useDispatch();

  const extent: Extent = getExtent(map);
  useEffect(() => {
    // For now, assume that if we have layer data, we don't need to refetch. This could change down the line if we
    // want to dynamically re-fetch data based on changing map bounds.
    // Only fetch once we actually know the extent
    const [minX, , maxX] = extent;
    if (
      selectedDate &&
      (!data || date !== selectedDate) &&
      minX !== 0 &&
      maxX !== 0
    ) {
      dispatch(loadLayerData({ layer, extent, date: selectedDate }));
    }
  }, [dispatch, layer, extent, data, selectedDate, date]);

  if (!data) {
    return selectedDate ? null : (
      <div className={classes.message}>
        <div className={classes.messageContainer}>
          <h2>Select an available date to view data</h2>
        </div>
      </div>
    );
  }

  const { impactFeatures, boundaries } = data;
  const noMatchingDistricts = impactFeatures.features.length === 0;

  const fillPaint: FillPaint = {
    'fill-opacity': layer.opacity || 0.1,
    'fill-color': noMatchingDistricts
      ? 'gray'
      : {
          property: 'impactValue',
          stops: legendToStops(layer.legend),
        },
  };

  const hazardLayerDef = LayerDefinitions[layer.hazardLayer];
  const operation = layer.operation || 'median';
  const hazardTitle = `${hazardLayerDef.title} (${operation})`;
  const boundaryId = getBoundaryLayerSingleton().id;

  return (
    <GeoJSONLayer
      before={`layer-${boundaryId}-line`}
      id={`layer-${layer.id}`}
      data={noMatchingDistricts ? boundaries : impactFeatures}
      linePaint={linePaint}
      fillPaint={fillPaint}
      fillOnClick={(evt: any) => {
        const popupData = {
          [layer.title]: {
            data: get(evt.features[0], 'properties.impactValue', 'No Data'),
            coordinates: evt.lngLat,
          },
          [hazardTitle]: {
            data: getHazardData(evt, operation),
            coordinates: evt.lngLat,
          },
        };
        // by default add `impactValue` to the tooltip
        dispatch(addPopupData(popupData));
        // then add feature_info_props as extra fields to the tooltip
        dispatch(
          addPopupData(
            getFeatureInfoPropsData(layer.featureInfoProps || {}, evt),
          ),
        );
      }}
    />
  );
};

const styles = (theme: Theme) =>
  createStyles({
    message: {
      position: 'absolute',
      height: '100%',
      width: '100%',
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    messageContainer: {
      fontSize: 24,
      paddingLeft: theme.spacing(2),
      paddingRight: theme.spacing(2),
      backgroundColor: theme.palette.grey.A100,
      borderRadius: theme.spacing(2),
    },
  });

interface ComponentProps extends WithStyles<typeof styles> {
  layer: ImpactLayerProps;
}

export default withStyles(styles)(ImpactLayer);
