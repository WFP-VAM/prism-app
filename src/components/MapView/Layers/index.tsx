import React, { Fragment } from 'react';
import { Source, Layer } from 'react-mapbox-gl';

import { formatServerUri } from '../../../utils/server-utils';
import { LayersMap } from '../../../config/types';

const wmsCommonQuery = {
  version: '1.1.1',
  request: 'GetMap',
  format: 'image/png',
  transparent: true,
  exceptions: 'application/vnd.ogc.se_inimage',
  bboxsr: 3857,
  imagesr: 3857,
  width: 256,
  height: 256,
  srs: 'EPSG:3857',
  bbox: '{bbox-epsg-3857}',
};

function Layers({ layers, selectedDate }: LayersProps) {
  return (
    <>
      {layers &&
        layers
          .valueSeq()
          .toJS()
          .map(({ id, serverUri, opacity }) => (
            <Fragment key={id}>
              <Source
                id={`source-${id}`}
                tileJsonSource={{
                  type: 'raster',
                  tiles: [
                    formatServerUri(serverUri, wmsCommonQuery, selectedDate),
                  ],
                  tileSize: 256,
                }}
              />

              <Layer
                type="raster"
                id={`layer-${id}`}
                sourceId={`source-${id}`}
                paint={{ 'raster-opacity': opacity }}
              />
            </Fragment>
          ))}
    </>
  );
}

export interface LayersProps {
  layers: LayersMap;
  selectedDate?: number;
}

export default Layers;
