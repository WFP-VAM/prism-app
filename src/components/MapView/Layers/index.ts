import WMSLayer from './WMSLayer';
import NSOLayer from './NSOLayer';
import GroundstationLayer from './GroundstationLayer';
import BoundaryLayer from './BoundaryLayer';
import ImpactLayer from './ImpactLayer';

export { ImpactLayer, WMSLayer, NSOLayer, GroundstationLayer, BoundaryLayer };
