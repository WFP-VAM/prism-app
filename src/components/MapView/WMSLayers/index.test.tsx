import React from 'react';
import { Map } from 'immutable';
import { render } from '@testing-library/react';

import Layers from '.';

test('renders as expected', () => {
  // TODO: Mock layers
  const { container } = render(<Layers layers={Map()} />);
  expect(container).toMatchSnapshot();
});
