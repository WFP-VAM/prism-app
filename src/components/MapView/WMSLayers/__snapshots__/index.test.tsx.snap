// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `<div />`;
